### PR TITLE
[WIP] keep original pod ip on a dummy interface 

### LINKED ISSUE #2076: VMIs with delegated IPs are stopped by the kubelet if kubelet is restarted
**Is this a BUG REPORT or FEATURE REQUEST?**:

> Uncomment only one, leave it on its own line:
>
/kind bug
> /kind enhancement

**What happened**:

When connecting the VMI via `bridge` mode and IP delegation to the pod network, CNI status checks fail and as a consequence the kubelet stops all containers of that pod when the kubelet is restarted.

I checked it on k8s 1.10.4 and k8s 1.13.3. I *think* that the hint in the logs is this:

```
Mar 04 09:28:43 node01 kubelet[29530]: W0304 09:28:43.167534   29530 docker_sandbox.go:384] failed to read pod IP from plugin/docker: NetworkPlugin cni failed on the status hook for pod "virt-launcher-vmi-nocloud-v4j2m_default": Unexpected address output
```

I did not try it with ovn-cni, crio, just flannel.

**What you expected to happen**:

The VMIs should not be killed by the kubelet.

**How to reproduce it (as minimally and precisely as possible)**:

```bash
$ cluster/kubectl.sh create -f cluster/examples/vmi-nocloud.yaml
$ cluster/kubectl.sh create -f cluster/examples/vmi-slirp.yaml
$ cluster/kubectl.sh get vmis
NAME          AGE   PHASE     IP            NODENAME
vmi-nocloud   15s   Running   10.244.0.32   node01
vmi-slirp     1m    Running   10.244.0.31   node01
$ cluster/cli.sh ssh node01
$ sudo systemctl restart kubelet
$ cluster/kubectl.sh get vmis
NAME          AGE   PHASE     IP            NODENAME
vmi-nocloud   34s   Failed    10.244.0.32   node01
vmi-slirp     1m    Running   10.244.0.31   node01
```

**Anything else we need to know?**:

**Environment**:
- KubeVirt version (use `virtctl version`):
- Kubernetes version (use `kubectl version`):
- VM or VMI specifications:
- Cloud provider or hardware configuration:
- OS (e.g. from /etc/os-release):
- Kernel (e.g. `uname -a`):
- Install tools:
- Others:

### PR DESCRIPTION
**Pull request by phoracek**
**What this PR does / why we need it**:

When kubelet is restarted, it goes over all pods, peeks in their
respective network namespaces and tries to detect their IP in order
to update their Pod objects.

However, with the bridge binding mode, we move this IP address to VM
and the pod interface is left IP-less. That causes kubelet's IP check
to fail and results in pod removal.

With this patch, we use a dummy interface named after the pod interface
to keep the pod IP address. When kubelet performs the check, it sees
interface with expectected name (eth0) and reports its IP address on
Pod object.

Fixes #2076

**Release note**:

```release-note
fix pod default network IP reporting: network bridge binding method keeps original pod address on dummy interface
```

### CODE REVIEW

**Comment by phoracek** [comment]
Failed on unrelated issues downloading an image :/

**Comment by phoracek** [comment]
@cynepco3hahue @rmohr this PR fails on (I believe) unrelated issues, could you suggest me what should I do? I know we switched to bazel and there could be some issues related to it, so I'd like to know whether I should try to retriggeer, wait or rebase. Thanks

**Comment by phoracek** [comment]
@SchSeba @booxter @rmohr @davidvossel could you please take a look?

**Comment by vladikr** [comment]
@phoracek The direction looks good to me, but this PR misses any kind of tests...
Are you planning to add any?

**Comment by phoracek** [comment]
@vladikr good point. @rmohr do you plan to have a test in #2068 where you actually restart kubelet, or do you just test the device plugin separately? Do you know about a way to trigger kubelet's check without restarting it?

**Comment by phoracek** [comment]
Also, I need to change tests so they expect that name of the interface changed.

**Comment by rmohr** [comment]
> @vladikr good point. @rmohr do you plan to have a test in #2068 where you actually restart kubelet, or do you just test the device plugin separately? Do you know about a way to trigger kubelet's check without restarting it?

I was happy that I didn't have to. In theory you can start a privileged pod and simply kill the kubelet (that's also how we kill virt-handler in tests). The pod will not immediately report a success, since the connection to the node will be lost for some time, but after the kubelet is back the pod should go to success and you can then check if te vmis are still alive.

**Comment by phoracek** [comment]
@rmohr thanks! Sounds easy enough

**Comment by rmohr** [comment]
ci test please

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 1]
```diff
@@ -344,6 +349,37 @@ func (b *BridgePodInterface) createBridge() error {
...
+	// Rename pod interface to free the original name for a new dummy interface
+	err := Handler.LinkSetName(b.podNicLink, newPodInterfaceName)
+	if err != nil {
+		log.Log.Reason(err).Errorf("failed to rename interface : %s", b.podInterfaceName)
+		return err
+	}
+	b.podInterfaceName = newPodInterfaceName
+
+	// Create a dummy interface named after the original interface
+	dummy := &netlink.Dummy{LinkAttrs: netlink.LinkAttrs{Name: originalPodInterfaceName}}
```
Comment (rmohr) [inline pkg/virt-launcher/virtwrap/network/podinterface.go]: I think there were some concerns on what would happen in cni plugins if we change the interface type. I guess the risk there that something bad is happening, but they could also fail now because of the missing ip. I guess the risk is pretty much the same with this change. We change one property (name) to something unexpected but at the same time ensure that another property is now the same value like before (ip).

**Comment by rmohr** [comment]
> Failed on unrelated issues downloading an image :/

Yep the mirror of the image is really an issue.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 2]
```diff
@@ -344,6 +349,37 @@ func (b *BridgePodInterface) createBridge() error {
 	return nil
 }
 
+func (b *BridgePodInterface) switchPodInterfaceWithDummy() error {
+	originalPodInterfaceName := b.podInterfaceName
+	newPodInterfaceName := fmt.Sprintf("%s-nic", originalPodInterfaceName)
+
+	// Rename pod interface to free the original name for a new dummy interface
+	err := Handler.LinkSetName(b.podNicLink, newPodInterfaceName)
```
Comment (davidvossel) [inline pkg/virt-launcher/virtwrap/network/podinterface.go]: is this function idempotent? meaning if we get a random failure somewhere in here and the function is tried again, can it succeed regardless of what occurred the previous execution?

For example, should we be checking to see if things like the dummy nic already exists and the rename occurred before doing the rename/swap. Otherwise it looks like this might fail if it gets called twice.

Comment (rmohr) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: Yep. Let's verify that this is idempotent. :+1:

**Comment by fabiand** [comment]
@phoracek I guess you already have the tests in your local branch, do you mind to push them?

CI is a crucial part of our KubeVirt work.

**Comment by davidvossel** [comment]
i'd like to see this picked back up. Even though we're de-emphasizing bridge mode, it's still an important feature we need to support.

**Comment by phoracek** [comment]
All right, lemme rebase and try again.

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 2 above)
Comment (phoracek) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: It is not. Will fix it.

File: pkg/virt-launcher/virtwrap/network/podinterface.go (hunk 1 above)
Comment (phoracek) [reply pkg/virt-launcher/virtwrap/network/podinterface.go]: Let's pretend we are improving it by promoting to the higher level API - L3 instead of L2.

File: pkg/virt-launcher/virtwrap/network/podinterface.go [hunk 3]
```diff
@@ -344,6 +349,37 @@ func (b *BridgePodInterface) createBridge() error {
 	return nil
 }
 
+func (b *BridgePodInterface) switchPodInterfaceWithDummy() error {
+	originalPodInterfaceName := b.podInterfaceName
+	newPodInterfaceName := fmt.Sprintf("%s-nic", originalPodInterfaceName)
```
Comment (booxter) [inline pkg/virt-launcher/virtwrap/network/podinterface.go]: This is probably a corner case, but there is a limit as to the length of interfaces in Linux. If you append four more characters, in a degenerate case you may end up with a name that is not accepted by the kernel. Are there any standards in Kubernetes as to the maximal length of the pod interface name?

File: tests/vmi_networking_test.go [hunk 4]
```diff
@@ -100,7 +100,7 @@ var _ = Describe("[rfe_id:694][crit:medium][vendor:cnv-qe@redhat.com][level:comp
 	}
 
 	checkLearningState := func(vmi *v1.VirtualMachineInstance, expectedValue string) {
```
Comment (booxter) [inline tests/vmi_networking_test.go]: it would be nice to have a test that would validate the IP address on a dummy

**Comment by rmohr** [comment]
ping @phoracek what is the state on this one? We discussed finishing this soon already multiple times the last months.

**Comment by phoracek** [comment]
@rmohr the scenario is always the same, when I find some time to dedicate to this PR, I run tests a few times, get frustrated how long it takes and how flaky it is and give up. Plus there is a new problem with this one. Since we switched our functional tests to `masquerade` there is way less test coverage for `bridge` binding, therefore any change on the code is way more risky. Let me try again. No promises.

**Comment by rmohr** [comment]
/reopen

Still relevant

**Comment by rthallisey** [comment]
New PR: https://github.com/kubevirt/kubevirt/pull/4552
